mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/SubscriptionStateHandlerImpl.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.mantisrx.server.worker;
+
+import io.mantisrx.server.master.client.MantisMasterGateway;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Subscription;
+
+class SubscriptionStateHandlerImpl implements SinkSubscriptionStateHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(SubscriptionStateHandlerImpl.class);
+    private final AtomicReference<ScheduledFuture> timedOutExitFutureRef = new AtomicReference<>();
+    private final String jobId;
+    private final MantisMasterGateway masterClientApi;
+    private ScheduledThreadPoolExecutor executor;
+    private final long subscriptionTimeoutSecs;
+    private final long minRuntimeSecs;
+    private long startedAt = System.currentTimeMillis();
+
+    SubscriptionStateHandlerImpl(String jobId, MantisMasterGateway masterClientApi, long subscriptionTimeoutSecs, long minRuntimeSecs) {
+        this.jobId = jobId;
+        this.masterClientApi = masterClientApi;
+        this.subscriptionTimeoutSecs = subscriptionTimeoutSecs;
+        this.minRuntimeSecs = minRuntimeSecs;
+        executor = this.subscriptionTimeoutSecs > 0L ? new ScheduledThreadPoolExecutor(1) : null;
```

Comment:
this might be a good candidate for optional IMO.

Reply:
This is existing code. I just deleted the file that this was supposed to replace, and you should be able to see the changes I made to this file now.

Reply:
I have completely refactored the code. Can you guys take a look now? It should be way simpler than the original version.